### PR TITLE
Add latest updates from FTD Ansible downstream repository.

### LINKED ISSUE #79: Cisco ftd_configuration module not idempotent
##### SUMMARY

Cisco ftd_configuration ansible module is not idempotent when tried with multiple references of the same object in array.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

ftd_configuration

##### ANSIBLE VERSION

```paste below
2.7.5
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

mac os

##### STEPS TO REPRODUCE

```yaml
- hosts: ftd
  connection: httpapi
  tasks:
- name: Create URL Object
  ftd_configuration:
    operation: upsertURLObject
    data:
      name: GoogleURL
      description: URL for Google
      url: www.google.com
      type: urlobject
      register_as: url_obj

- name: Find Security Intelligence URL Policy
  ftd_configuration:
    operation: getSecurityIntelligenceURLPolicyList
    register_as: policies

- name: Setup SecurityIntelligenceURLPolicy
  ftd_configuration:
    operation: editSecurityIntelligenceURLPolicy
   data:
     id: "{{ policies[0]['id'] }}"
    name: "{{ policies[0]['name'] }}"
    version: "{{ policies[0]['version'] }}"
    whitelist:
    - "{{ policies[0]['whitelist'] + [url_obj] }}"
    type: securityintelligenceurlpolicy
    path_params:
      objId: "{{ policies[0]['id'] }}" 
```

##### EXPECTED RESULTS

Play should be idempotent in consecutive runs.

##### ACTUAL RESULTS

Play is not idempotent in consecutive runs and shows as `changed=True`, every times the play runs.

```paste below

```

### PR DESCRIPTION
**Pull request by 119Vik**
##### SUMMARY
In the scope of this PR following improvement were implemented for FTD related modules
 - add a better implementation of the upsert operation;
 - add API version lookup functionality;
 - add a filter which removes duplicated references from the list of references, fixes [issue](https://github.com/CiscoDevNet/FTDAnsible/issues/79).

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
plugins.httpapi.ftd
modules.network.ftd.ftd_configuration

### CODE REVIEW

**Comment by kbreit** [comment]
+label cisco

**Comment by 119Vik** [comment]
Hi @justjais is there any feedback regarding this PR?

File: lib/ansible/module_utils/network/ftd/configuration.py [hunk 1]
```diff
@@ -371,9 +381,12 @@ def edit_object(self, operation_name, params):
         return self.send_general_request(operation_name, params)
 
     def send_general_request(self, operation_name, params):
+        def stop_if_check_mode():
```
Comment (justjais) [inline lib/ansible/module_utils/network/ftd/configuration.py:384]: @119Vik can you please comment the reason you have changed the implementation as this.

File: lib/ansible/module_utils/network/ftd/configuration.py [hunk 2]
```diff
@@ -31,9 +31,19 @@
 DUPLICATE_NAME_ERROR_MESSAGE = "Validation failed due to a duplicate name"
 
 MULTIPLE_DUPLICATES_FOUND_ERROR = (
-    "Cannot add a new object. An object(s) with the same attributes exists."
-    "Multiple objects returned according to filters being specified. "
-    "Please specify more specific filters which can find exact object that caused duplication error")
+    "Multiple objects matching specified filters are found. "
+    "Please, define filters more precisely to match one object exactly."
+)
+DUPLICATE_ERROR = (
```
Comment (justjais) [inline lib/ansible/module_utils/network/ftd/configuration.py:37]: @119Vik wanted to verify if these exceptions are FTD exceptions, or Ansible specific.

File: lib/ansible/module_utils/network/ftd/fdm_swagger_client.py [hunk 3]
```diff
@@ -506,7 +516,7 @@ def _is_enum(self, model):
         return self._is_string_type(model) and PropName.ENUM in model
 
     def _check_enum(self, status, model, value, path):
```
Comment (justjais) [inline lib/ansible/module_utils/network/ftd/fdm_swagger_client.py]: can we change the name of `value` to `data` if feasible

**Review by justjais — COMMENTED** [review]
Thanks @119Vik for the PR, I have commented few cosmetic review comments once resolved shall merge the PR.

File: lib/ansible/module_utils/network/ftd/configuration.py (hunk 2 above)
Comment (119Vik) [reply lib/ansible/module_utils/network/ftd/configuration.py:37]: @justjais these exceptions are ansible module specific.

File: lib/ansible/module_utils/network/ftd/configuration.py (hunk 1 above)
Comment (119Vik) [reply lib/ansible/module_utils/network/ftd/configuration.py:384]: @justjais this change was done to make the code a bit more self-explaining. Name of the method should describe the purpose of the `if` and `raise` statements.

File: lib/ansible/module_utils/network/ftd/fdm_swagger_client.py (hunk 3 above)
Comment (119Vik) [reply lib/ansible/module_utils/network/ftd/fdm_swagger_client.py]: @justjais good catch. fixed.

**Review by justjais — APPROVED** [review]

**Pull request merged**